### PR TITLE
match wildcards values

### LINKED ISSUE #125: Allow wildcards in Client Domain and callback url fields
Currently, comma separated values for client Domain and callback URL fields are allowed. Need to add support for wildcards in these fields like

## What?
Allow wildcards values, like `https://*.codingblocks.com`. This can be implemented using regexp or take regexp as input from the user. @championswimmer if you can describe more about this.

## Why ?
We use netlify for building and hosting our frontend clients. However, the deploy previews generated cannot use oneauth because each deploy preview is hosted on a subdomain of its own.

### PR DESCRIPTION
**Pull request by architkshk**
fixes #125 . @abhishek97 Please check.

### CODE REVIEW

**Comment by championswimmer** [comment]
@abhishek97 this would do it ?

File: src/oauth/oauthserver.js [hunk 1]
```diff
@@ -79,7 +79,9 @@ server.exchange(oauth.exchange.code(
             }
             let callbackMatch = false
             for (url of client.callbackURL) {
-                if (redirectURI.startsWith(url)) callbackMatch = true
+                const exp = "[^\s]+";
```
Comment (abhishek97) [inline src/oauth/oauthserver.js]: Why not `[^\s]*` ? wildcards and be replaced by empty strings also.

**Comment by abhishek97** [comment]
@championswimmer This would do for matching, however, I am not sure 

- Is this the only place where we match `redirectUrl` with the client's saved `callbackUrl` CSV? We also have a `domain` field. I think that would also have a "check" of its own.

- Probably we need a similar treatment here too: https://github.com/architkshk/oneauth/blob/2b1dea7b238ef60423424afe1bf4f0b89fa9142f/src/oauth/oauthserver.js#L130 

- This custom matching logic can be refactored out in utils. Might help in reusablility and unit testing.

**Comment by architkshk** [comment]
@abhishek97 
- I cannot find any check for domain field.
- Replaced the other place where we have to check callback URL.
- Moved the code to utils/wildcardmatch.js

**Comment by abhishek97** [comment]
@architkshk Have you tried it with a sample app?

@championswimmer We can merge this on staging and test it out.

**Comment by architkshk** [comment]
@abhishek97 I've tried the code to check for various cases though I did not check for any sample app.

**Comment by architkshk** [comment]
@abhishek97 I've made changes to the code and also tried my creating sample clients on localhost. It works fine. Please review.

**Comment by architkshk** [comment]
@abhishek97 @championswimmer any updates?

**Comment by architkshk** [comment]
@abhishek97 @championswimmer any updates?

**Comment by abhishek97** [comment]
@championswimmer Let's try this out? I can find no checks for domains either.

**Comment by championswimmer** [comment]
@abhishek97 cool, will try and merge today/tomorrow

File: src/oauth/oauthserver.js [hunk 2]
```diff
@@ -78,10 +79,7 @@ server.exchange(oauth.exchange.code(
             if (client.id !== grantCode.client.id) {
                 return done(null, false) //Wrong Client ID
             }
-            let callbackMatch = false
-            for (url of client.callbackURL) {
-                if (redirectURI.startsWith(url)) callbackMatch = true
-            }
+            let callbackMatch = matchURL(client.callbackURL,redirectURIredirectURI);
```
Comment (championswimmer) [inline src/oauth/oauthserver.js]: dafuq? why did you remove the for loop :|

**Comment by championswimmer** [comment]
reverted. the for loop removal fucked up the entire logic

**Comment by architkshk** [comment]
@championswimmer what test did it failed?

**Comment by championswimmer** [comment]
none of the explicit auths were working

File: src/oauth/oauthserver.js (hunk 2 above)
Comment (architkshk) [reply src/oauth/oauthserver.js]: figured out the error. it's a type, in line 82 I have written redirectURI 2 times. I am making changes and recreating a PR.

**Comment by architkshk** [comment]
Done #235  works fine now.

**Comment by architkshk** [comment]
Done #284  works fine now. @championswimmer
